usgpo/get-links.py

```diff
@@ -0,0 +1,187 @@
+            },
+        )
+        if response.status_code == 200:
+            output = response.json()
+
+            for record in output["packages"]:
+                package_queue.put(record)
+                pbar.update(1)
+
+            url = output["nextPage"]
+            offset_mark = None
+            # Prevent too many API requests in a short period of time
+            time.sleep(5)
+        else:
+            logger.error(
+                f"get_packages received status code {response.status_code} for query {url}"
+            )
+            break
+
+    package_queue.put(None)
+
+
+def get_file_links(api_key, package):
+    package_id = package["packageId"]
+    response = api_query(
+        f"https://api.govinfo.gov/packages/{package_id}/summary",
+        headers={"accept": "application/json"},
+        params={"api_key": args.api_key},
+    )
+    if response.status_code == 200:
+        output = response.json()
+        return output.get("download")
+    return None
+
+
+def get_package_metadata(api_key, package_queue, metadata_queue):
+    pbar = tqdm(desc="Consumer")
+    while True:
+        package = package_queue.get()
+        if package is None:
+            package_queue.put(None)
+            metadata_queue.put(None)
+            break
+
+        record = {
+            "title": package.get("title"),
+            "package_id": package.get("packageId"),
+            "date": package.get("dateIssued"),
+            "category": package.get("category"),
+            "author": package.get("governmentAuthor1"),
+            "publisher": package.get("publisher"),
+            "links": get_file_links(api_key, package),
+        }
+        metadata_queue.put(record)
+        pbar.update(1)
+
+
+def write_metadata(output_dir, metadata_queue):
+    with jsonlines.open(os.path.join(output_dir, "links.jsonl"), mode="w") as writer:
+        pbar = tqdm(desc="Writer")
+        while True:
+            metadata = metadata_queue.get()
+            if metadata is None:
+                metadata_queue.task_done()
+                break
+
+            writer.write(metadata)
+            pbar.update(1)
+
+
+def main(args):
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    package_queue = queue.Queue()
+    metadata_queue = queue.Queue()
+
+    with ThreadPoolExecutor(max_workers=args.workers + 2) as executor:
+        # One thread for getting each package (i.e. file) from the specified collections
+        executor.submit(
+            get_packages, args.api_key, args.collections, args.start_date, package_queue
+        )
+
+        # `args.workers` threads for getting package metadata
+        for _ in range(args.workers):
+            executor.submit(
+                get_package_metadata, args.api_key, package_queue, metadata_queue
+            )
+
+        # One thread for writing out the package metadata to disk
+        executor.submit(write_metadata, args.output_dir, metadata_queue)
```

Comment:
I think you have a race condition in this parallelism. Let the package_queue have 2 packages followed the `None` sentinel, and there are 2 two workers. 1 worker gets the links quickly while the second worker gets a `429` and sleeps for an hour. worker 1 writes the package metadata to the metadata_queue and gets the next item from the package queue, the `None`, and writes it to the metadata queue. This means the writer will see the `None` before the metadata for package 2, the writer will close, and the package 2 metadata will be lost.

Reply:
I think the way people generally do this is each worker needs to add a sentinel to the writer queue and the write only stops when it sees #workers sentinels